File: src/main/java/me/grabsky/claims/claims/Claim.java
package me.grabsky.claims.claims;

import com.sk89q.worldedit.bukkit.BukkitAdapter;
import com.sk89q.worldguard.protection.flags.Flags;
import com.sk89q.worldguard.protection.regions.ProtectedRegion;
import me.grabsky.claims.Claims;
import me.grabsky.claims.configuration.ClaimsConfig;
import me.grabsky.claims.flags.ExtraFlags;
import org.bukkit.Location;

import java.util.Set;
import java.util.UUID;

public class Claim {
    private final static ClaimManager manager = Claims.getInstance().getClaimManager();

    private final String id;
    private final UUID owner;
    private ProtectedRegion wgRegion;

    public Claim(String id, UUID owner, ProtectedRegion wgRegion) {
        this.id = id;
        this.wgRegion = wgRegion;
        this.owner = owner;
    }

    protected void update(final ProtectedRegion wgRegion) {
        this.wgRegion = wgRegion;
    }

    public String getId() {
        return id;
    }

    public ProtectedRegion getWGRegion() {
        return wgRegion;
    }

    public UUID getOwner() {
        return owner;
    }

    public int getLevel() {
        final Integer value = wgRegion.getFlag(ExtraFlags.CLAIM_LEVEL);
        return (value == null) ? 0 : value;
    }

    // This shouldn't be null unless manually deleted
    public Location getCenter() {
        return BukkitAdapter.adapt(wgRegion.getFlag(ExtraFlags.CLAIM_CENTER));
    }

    // This shouldn't be null unless manually deleted
    public Location getHome() {
        return BukkitAdapter.adapt(wgRegion.getFlag(Flags.TELE_LOC));
    }

    public boolean setHome(Location location) {
        final com.sk89q.worldedit.util.Location loc = BukkitAdapter.adapt(location);
        if (wgRegion.contains(loc.getBlockX(), loc.getBlockY(), loc.getBlockZ())) {
            wgRegion.setFlag(Flags.TELE_LOC, loc);
            return true;
        }
        return false;
    }

    public Set<UUID> getMembers() {
        return wgRegion.getMembers().getUniqueIds();
    }

    public boolean isMember(UUID uuid) {
        return wgRegion.getMembers().contains(uuid);
    }

    public boolean addMember(UUID uuid) {
        if (this.getMembers().size() < ClaimsConfig.MEMBERS_LIMIT) {
            wgRegion.getMembers().addPlayer(uuid);
            manager.getClaimPlayer(uuid).addRelative(this.getId());
            return true;
        }
        return false;
    }

    public boolean removeMember(UUID uuid) {
        if (this.getMembers().contains(uuid)) {
            wgRegion.getMembers().removePlayer(uuid);
            manager.getClaimPlayer(uuid).removeRelative(this.getId());
            return true;
        }
        return false;
    }

    /* Utility Methods */

    // 'Generates' region ID for specified location
    public static String createId(final org.bukkit.Location location) {
        return new StringBuilder().append(ClaimsConfig.REGION_PREFIX).append("x").append(location.getBlockX()).append("y").append(location.getBlockY()).append("z").append(location.getBlockZ()).toString();
    }
}

File: src/main/kotlin/claims/plugin/claims/Claim.kt
package claims.plugin.claims

import com.sk89q.worldedit.bukkit.BukkitAdapter
import com.sk89q.worldguard.protection.flags.Flags
import com.sk89q.worldguard.protection.regions.ProtectedRegion
import java.util.*

object Claims {
    fun id(location: org.bukkit.Location): String {
        return buildString {
            append("PREFIX")
            append("x${location.blockX}")
            append("y${location.blockX}")
            append("z${location.blockX}")
        }
    }
}

class Claim(val id: String, val owner: UUID, var region: ProtectedRegion) {

    fun getHome(): org.bukkit.Location? {
        return BukkitAdapter.adapt(region.getFlag(Flags.TELE_LOC))
    }

    fun setHome(location: org.bukkit.Location): Boolean {
        // Returning 'false' if provided location is not inside region
        if (region.contains(location.blockX, location.blockY, location.blockZ) == false) return false
        // Creating WorldEdit's Location object from org.bukkit.Location
        val notBukkitLocation = BukkitAdapter.adapt(location)
        // Setting 'teleport-location' flag
        region.setFlag(Flags.TELE_LOC, notBukkitLocation)
        // Returning 'true' after operation completion
        return true
    }

    fun getMembers(): Set<UUID> {
        return region.members.uniqueIds
    }

    fun isMember(uuid: UUID): Boolean {
        return region.members.uniqueIds.contains(uuid)
    }

    fun addMember(uuid: UUID): Boolean {
        // Returning 'false' if region reached members limit
        if (this.getMembers().size >= 10) return false
        // Adding player to the region
        region.members.addPlayer(uuid)
        // Returning 'true' after operation completion
        return true
    }

    fun removeMember(uuid: UUID): Boolean {
        // Returning 'false' if provided player is not a claim member
        if (this.isMember(uuid) == false) return false
        // Removing player from the region
        region.members.removePlayer(uuid)
        // Returning 'true' after operation completion
        return true
    }

    // TO-DO...
    fun upgrade(): Boolean {
        return true
    }
}